src/cluster/placement/algo/mirrored.go

```diff
@@ -225,6 +231,304 @@ func (a mirroredAlgorithm) ReplaceInstances(
+				return false
+			}
+
+			if leavingNode == "" {
+				// first iteration
+				leavingNode = s.SourceID()
+				continue
+			}
+
+			if leavingNode != s.SourceID() {
+				// came from a different node
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+type leavingShardKey struct{
+	LeavingShardSetID uint32
+	ShardID       uint32
+}
+
+func (a mirroredAlgorithm) markReplacingInstancesAvailable(p placement.Placement, leavingInstanceIDs []string) (placement.Placement, error) {
+	helper, err := newReplaceMarkAvailableHelper(a, p)
+	if err != nil {
+	    return nil, err
+	}
+
+	return helper.MarkLeavingInstancesAvailable(p, leavingInstanceIDs)
+}
+
+type replaceMarkAvailableHelper struct{
+	algo                              mirroredAlgorithm
+	leavingShardSetIDToOwningShardSet map[leavingShardKey]uint32
+	shardSetIDToOwnerIDs              map[uint32][]string
+}
+
+func newReplaceMarkAvailableHelper(
+	algo mirroredAlgorithm,
+	p placement.Placement,
+) (replaceMarkAvailableHelper, error) {
+	leavingShardIDToOwningInst := make(map[leavingShardKey]uint32)
+	for _, inst := range p.Instances() {
+		for _, s := range inst.Shards().All() {
+			if s.State() != shard.Initializing {
+				continue
+			}
+			sourceID := s.SourceID()
+			if sourceID == "" {
+				// this is true for initial placements
+				continue
+			}
+
+			// determine the source *shard set*
+			sourceInst, ok := p.Instance(sourceID)
+			if !ok {
+				return replaceMarkAvailableHelper{}, fmt.Errorf(
+					"source instance %q not in placement for shard %d on instance %q",
+					sourceID,
+					s.ID(),
+					inst.ID(),
+					)
+			}
+
+			leavingShardIDToOwningInst[leavingShardKey{
+				ShardID:           s.ID(),
+				LeavingShardSetID: sourceInst.ShardSetID(),
+			}] = inst.ShardSetID()
+		}
+	}
+
+	shardSetIDToOwners := make(map[uint32][]string)
+	for _, inst := range p.Instances() {
+		shardSetIDToOwners[inst.ShardSetID()] = append(shardSetIDToOwners[inst.ShardSetID()], inst.ID())
+	}
+	return replaceMarkAvailableHelper{
+		algo:                              algo,
+		leavingShardSetIDToOwningShardSet: leavingShardIDToOwningInst,
+		shardSetIDToOwnerIDs:              shardSetIDToOwners,
+	}, nil
+}
+
+func (h replaceMarkAvailableHelper) MarkLeavingInstancesAvailable(
+	p placement.Placement,
+	leavingInstanceIDs []string,
+) (placement.Placement, error) {
+	// lookup the entire shardset
+	// amainsd: what if we get the same shard set multiple times (e.g. multi node replace in same shardset)? Could dedupe here.
+	var err error
+	for _, leavingInstID := range leavingInstanceIDs {
+		leavingInst, ok := p.Instance(leavingInstID)
+		if !ok {
+			return nil, fmt.Errorf("")
+		}
+
+		shardSetOwners, ok := h.shardSetIDToOwnerIDs[leavingInst.ShardSetID()]
+		if !ok {
+			return nil, fmt.Errorf(
+				"shard set %d isn't owned by anything in the placement; placement may be invalid",
+				leavingInst.ShardSetID(),
+			)
+		}
+		p, err = h.markReplacePairAvailable(p, shardSetOwners)
```

Comment:
Before calling this, shouldn't we ensure that `len(shardSetOwners) >= 2`? Or maybe even ensure it earlier when building that map?